hack/generator/spec/defaulter-validator.md

```diff
@@ -0,0 +1,100 @@
+# Custom validation and defaulting for code generated resources
+
+## Reasoning
+`controller-runtime` defines `admission.Defaulter` and `admission.Validator`. These interfaces only give you a single `Default` or `ValidateX` method, which means that all validation/defaulting needs to be done in that method. I think we're going to quickly run into situations where we want custom (handcrafted) validations or defaults for a particular resource and we're not going to want to teach the code generator about these.
+
+## Suggestion
+Structure our autogenerated webhooks such that they're a pipeline. Add the ability to "take control" of this process by implementing an interface.
+
+Implement one (or both) of these interfaces to take control of validation/defaulting by hand
+```
+type Validator interface {
+	CreateValidations() []func() error
+	UpdateValidations() []func(old runtime.Object) error
+	DeleteValidations() []func() error
+}
+
+type Defaulter interface {
+	Defaults() []func()
+}
+```
+
+Sample `Default`:
+```
+// +kubebuilder:webhook:path=/mutate-microsoft-storage-infra-azure-com-v1alpha1api20190401-storageaccount,mutating=true,sideEffects=None,matchPolicy=Exact,failurePolicy=fail,groups=microsoft.storage.infra.azure.com,resources=storageaccounts,verbs=create;update,versions=v1alpha1api20190401,name=default.v1alpha1api20190401.storageaccounts.microsoft.storage.infra.azure.com,admissionReviewVersions=v1beta1
+
+var _ admission.Defaulter = &StorageAccount{}
+
+// Default defaults the Azure name of the resource to the Kubernetes name
+func (storageAccount *StorageAccount) Default() {
+	defaults := storageAccount.defaults()
+	var temp runtime.Object = storageAccount
+	if runtimeDefaulter, ok := temp.(genruntime.Defaulter); ok {
+		defaults = runtimeDefaulter.Defaults()
```

Comment:
Wouldn't we want the collected default funcs to be added with the base one? Allowing users to override azure name defaulting (or more likely forget to do it in a default func they provide themselves) seems dangerous.

Reply:
I was thinking that the implementer could choose to call this in their `genruntime.Defauler` implementation if they wanted the existing defaults, otherwise there's no way to bypass an autogenerated default... but maybe we dont want that anyway. We could start out without the ability to do that I guess and then just add it if we ever needed it - fair enough.

Reply:
Ah - in the example it doesn't look like the base implementation would be available to call so if someone provided their own implementation of Defaulter they'd need to rewrite the base one, but you're saying it would be public. In that case I think it's fine - definitely more flexibility if they can choose whether it happens.

Reply:
It doesn't actually have to be public does it? Since the override has to come from inside this package?